src/ctap/pin_protocol_v1.rs

```diff
@@ -0,0 +1,1228 @@
+    // different behavior when there are non-zero bytes after zero bytes.
+    // This implementation consistently ignores those degenerate cases.
+    Some(
+        blocks
+            .iter()
+            .flatten()
+            .cloned()
+            .take_while(|&c| c != 0)
+            .collect::<Vec<u8>>(),
+    )
+}
+
+/// Stores the encrypted new PIN in the persistent storage, if it satisfies the
+/// PIN policy. The PIN is decrypted and stripped from its padding. Next, the
+/// length of the PIN is checked to fulfill policy requirements. Last, the PIN
+/// is hashed, truncated to 16 bytes and persistently stored.
+fn check_and_store_new_pin(
+    persistent_store: &mut PersistentStore,
+    aes_dec_key: &crypto::aes256::DecryptionKey,
+    new_pin_enc: Vec<u8>,
+) -> Result<(), Ctap2StatusCode> {
+    let pin = decrypt_pin(aes_dec_key, new_pin_enc)
+        .ok_or(Ctap2StatusCode::CTAP2_ERR_PIN_POLICY_VIOLATION)?;
+
+    #[cfg(feature = "with_ctap2_1")]
+    let min_pin_length = persistent_store.min_pin_length() as usize;
+    #[cfg(not(feature = "with_ctap2_1"))]
+    let min_pin_length = 4;
+    if pin.len() < min_pin_length || pin.len() == PIN_PADDED_LENGTH {
+        // TODO(kaczmarczyck) check 4 code point minimum instead
+        return Err(Ctap2StatusCode::CTAP2_ERR_PIN_POLICY_VIOLATION);
+    }
+    let mut pin_hash = [0u8; 16];
+    pin_hash.copy_from_slice(&Sha256::hash(&pin[..])[..16]);
+    persistent_store.set_pin_hash(&pin_hash);
+    Ok(())
+}
+
+#[cfg(feature = "with_ctap2_1")]
+#[cfg_attr(test, derive(IntoEnumIterator))]
+// TODO remove when all variants are used
+#[allow(dead_code)]
+pub enum PinPermission {
+    // All variants should use integers with a single bit set.
+    MakeCredential = 0x01,
+    GetAssertion = 0x02,
+    CredentialManagement = 0x04,
+    BioEnrollment = 0x08,
+    PlatformConfiguration = 0x10,
+    AuthenticatorConfiguration = 0x20,
+}
+
+pub struct PinProtocolV1 {
+    key_agreement_key: crypto::ecdh::SecKey,
+    pin_uv_auth_token: [u8; PIN_TOKEN_LENGTH],
+    consecutive_pin_mismatches: u8,
+    #[cfg(feature = "with_ctap2_1")]
+    permissions: u8,
+    #[cfg(feature = "with_ctap2_1")]
+    permissions_rp_id: Option<String>,
+}
+
+impl PinProtocolV1 {
+    pub fn new(rng: &mut impl Rng256) -> PinProtocolV1 {
+        let key_agreement_key = crypto::ecdh::SecKey::gensk(rng);
+        let pin_uv_auth_token = rng.gen_uniform_u8x32();
+        PinProtocolV1 {
+            key_agreement_key,
+            pin_uv_auth_token,
+            consecutive_pin_mismatches: 0,
+            #[cfg(feature = "with_ctap2_1")]
+            permissions: 0,
+            #[cfg(feature = "with_ctap2_1")]
+            permissions_rp_id: None,
+        }
+    }
+
+    /// Decrypts the encrypted pin_hash and compares it to the stored pin_hash.
+    /// Resets or decreases the PIN retries, depending on success or failure.
+    fn verify_pin_hash_enc(
+        &mut self,
+        rng: &mut impl Rng256,
```

Comment:
I just noticed it now, but I find it strange that a verification function takes a rng as input. Could you document that the key agreement key is updated to a random key, why, and when?

Reply:
There is no reason in the specification for why the key has to be reset, so I don't want to speculate. It's most likely to prevent attacks on the private platform key-agreement key by cycling it. Everything else is added as a comment.